README.md
# [language-ansi-styles](https://atom.io/packages/language-ansi-styles)

Converts ANSI escape sequences into formatted text like in your terminal/shell.

![Demo](https://raw.githubusercontent.com/lloiser/language-ansi-styles/master/assets/sample.png)

Even basic images are possible:

![Smiley](https://raw.githubusercontent.com/lloiser/language-ansi-styles/master/assets/smiley.png)

# Options

* _Hide escape sequences_: hides all escape sequences and only keep the bare text with styles and color applied to it. <br /> __NOTE:__ Hiding characters in atom causes the editor to behave weirdly. The selection and cursor position is totally screwed and makes it pretty hard navigate within or select any text.

# Caveats

* the escape code for inverse `7` is currently not supported

* escape codes `38;2;r;g;b` and `48;2;r;g;b` are not supported

With these escape codes you can specify any rgb color (each color ranging from 0-255). Due to the nature of a grammar in atom it is not possible to create styles on demand as they occur. I could (in theory) create >16 million styles to capture every possible rgb color, but I doubt atom feels well afterwards...

* atom wraps after 500 characters, period.

atom v1.15.0 introduced a hard limit after how many characters a line automatically wraps to a line. This is actually not a big deal for a typical user. But keep it in mind before opening an issue.

* No themes at the moment

There is currently no possibility to define/select themes. For now you can override the colors in your own atom stylesheet.


CHANGELOG.md
## 0.1.0

* fix a breaking change in the atom grammar behavior

## 0.0.2

* improve readability by using more vibrant colors (by @azachar)

## 0.0.1

* It works!


spec/grammar-spec.js
'use babel'

/* eslint-env jasmine */
/* eslint no-multi-spaces: 0 */

import * as os from 'os'
import AnsiStylesGrammar from '../lib/ansi-styles-grammar'

describe('LanguageAnsiCodes', () => {
  let grammar = null

  beforeEach(() => {
    grammar = new AnsiStylesGrammar(atom.grammars)
  })

  function scopes (value, ...scopes) {
    return { value, scopes }
  }

  const testCases = [
    {
      title: 'simple',
      lines: [
        'colors: \x1b[30mblack\x1b[33myellow',
        'hello "Mr. \x1b[32mGreen\x1b[m" how are you?'
      ],
      expected: [
        [
          scopes('colors: ',   'ansi'),
          scopes('\x1b[30m',   'ansi', 'codes'),
          scopes('black',      'ansi', 'color.fg-0'),
          scopes('\x1b[33m',   'ansi', 'color.fg-0', 'codes'),
          scopes('yellow\r',   'ansi', 'color.fg-3')
        ],
        [
          scopes('hello "Mr. ',    'ansi', 'color.fg-3'),
          scopes('\x1b[32m',       'ansi', 'color.fg-3', 'codes'),
          scopes('Green',          'ansi', 'color.fg-2'),
          scopes('\x1b[m',         'ansi', 'color.fg-2', 'codes'),
          scopes('" how are you?', 'ansi')
        ]
      ]
    },

    {
      title: 'resets multiple on multiple lines',
      lines: [
        'one default \x1b[35m magenta \x1b[91m red',
        'two red \x1b[0m default',
        'three default \x1b[0m default',
        'four default \x1b[35m magenta \x1b[90m bright black',
        'five bright black \x1b[32;1m green bold \x1b[0;m default'
      ],
      expected: [
        [
          scopes('one default ', 'ansi'),
          scopes('\x1b[35m',     'ansi', 'codes'),
          scopes(' magenta ',    'ansi', 'color.fg-5'),
          scopes('\x1b[91m',     'ansi', 'color.fg-5', 'codes'),
          scopes(' red\r',       'ansi', 'color.fg-9')
        ],
        [
          scopes('two red ',     'ansi', 'color.fg-9'),
          scopes('\x1b[0m',      'ansi', 'color.fg-9', 'codes'),
          scopes(' default\r',   'ansi')
        ],
        [
          scopes('three default ', 'ansi'),
          scopes('\x1b[0m',        'ansi', 'codes'),
          scopes(' default\r',     'ansi')
        ],
        [
          scopes('four default ',   'ansi'),
          scopes('\x1b[35m',        'ansi', 'codes'),
          scopes(' magenta ',       'ansi', 'color.fg-5'),
          scopes('\x1b[90m',        'ansi', 'color.fg-5', 'codes'),
          scopes(' bright black\r', 'ansi', 'color.fg-8')
        ],
        [
          scopes('five bright black ', 'ansi', 'color.fg-8'),
          scopes('\x1b[32;1m',         'ansi', 'color.fg-8', 'codes'),
          scopes(' green bold ',       'ansi', 'color.fg-2', 'style.bold'),
          scopes('\x1b[0;m',           'ansi', 'color.fg-2', 'style.bold', 'codes'),
          scopes(' default',           'ansi')
        ]
      ]
    },

    {
      title: 'does not modify the simple text',
      lines: 'some text',
      expected: [scopes('some text', 'ansi')]
    },

    {
      title: 'renders foreground colors',
      lines: 'colors: \x1b[30mblack\x1b[37mwhite',
      expected: [
        scopes('colors: ', 'ansi'),
        scopes('\x1b[30m', 'ansi', 'codes'),
        scopes('black',    'ansi', 'color.fg-0'),
        scopes('\x1b[37m', 'ansi', 'color.fg-0', 'codes'),
        scopes('white',    'ansi', 'color.fg-7')
      ]
    },
    {
      title: 'renders bright foreground colors',
      lines: 'colors: \x1b[90mblack\x1b[97mwhite',
      expected: [
        scopes('colors: ', 'ansi'),
        scopes('\x1b[90m', 'ansi', 'codes'),
        scopes('black',    'ansi', 'color.fg-8'),
        scopes('\x1b[97m', 'ansi', 'color.fg-8', 'codes'),
        scopes('white',    'ansi', 'color.fg-15')
      ]
    },
    {
      title: 'renders background colors',
      lines: 'colors: \x1b[40mblack\x1b[47mwhite',
      expected: [
        scopes('colors: ', 'ansi'),
        scopes('\x1b[40m', 'ansi', 'codes'),
        scopes('black',    'ansi', 'color.bg-0'),
        scopes('\x1b[47m', 'ansi', 'color.bg-0', 'codes'),
        scopes('white',    'ansi', 'color.bg-7')
      ]
    },
    {
      title: 'renders bright background colors',
      lines: 'colors: \x1b[100mblack\x1b[107mwhite',
      expected: [
        scopes('colors: ',  'ansi'),
        scopes('\x1b[100m', 'ansi', 'codes'),
        scopes('black',     'ansi', 'color.bg-8'),
        scopes('\x1b[107m', 'ansi', 'color.bg-8', 'codes'),
        scopes('white',     'ansi', 'color.bg-15')
      ]
    },

    {
      title: 'renders extended fg colors',
      lines: 'colors: \x1b[38;5;196mextended fg red\x1b[0mdefault',
      expected: [
        scopes('colors: ',        'ansi'),
        scopes('\x1b[38;5;196m',  'ansi', 'codes'),
        scopes('extended fg red', 'ansi', 'color.fg-196'),
        scopes('\x1b[0m',         'ansi', 'color.fg-196', 'codes'),
        scopes('default',         'ansi')
      ]
    },
    {
      title: 'renders extended bg colors',
      lines: 'colors: \x1b[48;5;196mextended bg red\x1b[0mdefault',
      expected: [
        scopes('colors: ',        'ansi'),
        scopes('\x1b[48;5;196m',  'ansi', 'codes'),
        scopes('extended bg red', 'ansi', 'color.bg-196'),
        scopes('\x1b[0m',         'ansi', 'color.bg-196', 'codes'),
        scopes('default',         'ansi')
      ]
    },
    {
      title: 'renders extended fg colors over multiple lines',
      lines: [
        'colors: \x1b[38;5;196mextended fg red',
        'still fg red\x1b[0mdefault'
      ],
      expected: [
        [
          scopes('colors: ',          'ansi'),
          scopes('\x1b[38;5;196m',    'ansi', 'codes'),
          scopes('extended fg red\r', 'ansi', 'color.fg-196')
        ],
        [
          scopes('still fg red',      'ansi', 'color.fg-196'),
          scopes('\x1b[0m',           'ansi', 'color.fg-196', 'codes'),
          scopes('default',           'ansi')
        ]
      ]
    },
    {
      title: 'renders extended bg colors over multiple lines',
      lines: [
        'colors: \x1b[48;5;196mextended bg red',
        'still bg red\x1b[0mdefault'
      ],
      expected: [
        [
          scopes('colors: ',          'ansi'),
          scopes('\x1b[48;5;196m',    'ansi', 'codes'),
          scopes('extended bg red\r', 'ansi', 'color.bg-196')
        ],
        [
          scopes('still bg red',    'ansi', 'color.bg-196'),
          scopes('\x1b[0m',         'ansi', 'color.bg-196', 'codes'),
          scopes('default',         'ansi')
        ]
      ]
    },

    {
      title: 'renders faint',
      lines: 'faint: \x1b[2mthat\x1b[22m, default',
      expected: [
        scopes('faint: ',   'ansi'),
        scopes('\x1b[2m',   'ansi', 'codes'),
        scopes('that',      'ansi', 'style.faint'),
        scopes('\x1b[22m',  'ansi', 'style.faint', 'codes'),
        scopes(', default', 'ansi')
      ]
    },
    {
      title: 'renders strikethrough',
      lines: 'strike: \x1b[9mthat\x1b[29m, default',
      expected: [
        scopes('strike: ',  'ansi'),
        scopes('\x1b[9m',   'ansi', 'codes'),
        scopes('that',      'ansi', 'style.strike'),
        scopes('\x1b[29m',  'ansi', 'style.strike', 'codes'),
        scopes(', default', 'ansi')
      ]
    },
    {
      title: 'renders blink',
      lines: 'blink: \x1b[5mis so old school...\x1b[25m, default',
      expected: [
        scopes('blink: ',             'ansi'),
        scopes('\x1b[5m',             'ansi', 'codes'),
        scopes('is so old school...', 'ansi', 'style.blink'),
        scopes('\x1b[25m',            'ansi', 'style.blink', 'codes'),
        scopes(', default',           'ansi')
      ]
    },
    {
      title: 'renders underline',
      lines: 'underline: \x1b[4mme\x1b[24m, default',
      expected: [
        scopes('underline: ', 'ansi'),
        scopes('\x1b[4m',     'ansi', 'codes'),
        scopes('me',          'ansi', 'style.underline'),
        scopes('\x1b[24m',    'ansi', 'style.underline', 'codes'),
        scopes(', default',   'ansi')
      ]
    },
    {
      title: 'renders bold and resets it',
      lines: 'bold: \x1b[1mhero\x1b[21m, default',
      expected: [
        scopes('bold: ',    'ansi'),
        scopes('\x1b[1m',   'ansi', 'codes'),
        scopes('hero',      'ansi', 'style.bold'),
        scopes('\x1b[21m',  'ansi', 'style.bold', 'codes'),
        scopes(', default', 'ansi')
      ]
    },
    {
      title: 'renders italic',
      lines: 'italic: \x1b[3mfood\x1b[23m, default',
      expected: [
        scopes('italic: ',  'ansi'),
        scopes('\x1b[3m',   'ansi', 'codes'),
        scopes('food',      'ansi', 'style.italic'),
        scopes('\x1b[23m',  'ansi', 'style.italic', 'codes'),
        scopes(', default', 'ansi')
      ]
    },
    {
      title: 'hides text',
      lines: 'hidden: \x1b[8msecret\x1b[28m, default',
      expected: [
        scopes('hidden: ',  'ansi'),
        scopes('\x1b[8m',   'ansi', 'codes'),
        scopes('secret',    'ansi', 'style.hidden'),
        scopes('\x1b[28m',  'ansi', 'style.hidden', 'codes'),
        scopes(', default', 'ansi')
      ]
    },

    {
      title: 'handles a reset',
      lines: '\x1b[1mthis is bold\x1b[0m, but this isn\'t',
      expected: [
        scopes('\x1b[1m',           'ansi', 'codes'),
        scopes('this is bold',      'ansi', 'style.bold'),
        scopes('\x1b[0m',           'ansi', 'style.bold', 'codes'),
        scopes(', but this isn\'t', 'ansi')
      ]
    },
    {
      title: 'handles a reset with implicit 0',
      lines: '\x1b[1mthis is bold\x1b[m, but this isn\'t',
      expected: [
        scopes('\x1b[1m',           'ansi', 'codes'),
        scopes('this is bold',      'ansi', 'style.bold'),
        scopes('\x1b[m',            'ansi', 'style.bold', 'codes'),
        scopes(', but this isn\'t', 'ansi')
      ]
    },

    {
      title: 'handles multiple styles, fg and bg color',
      lines: 'normal, \x1b[1mbold, \x1b[4munderline, \x1b[3mitalic, \x1b[31mred fg, \x1b[44mblue bg\x1b[0m, normal',
      expected: [
        scopes('normal, ',    'ansi'),
        scopes('\x1b[1m',     'ansi', 'codes'),
        scopes('bold, ',      'ansi', 'style.bold'),
        scopes('\x1b[4m',     'ansi', 'style.bold', 'codes'),
        scopes('underline, ', 'ansi', 'style.bold', 'style.underline'),
        scopes('\x1b[3m',     'ansi', 'style.bold', 'style.underline', 'codes'),
        scopes('italic, ',    'ansi', 'style.bold', 'style.underline', 'style.italic'),
        scopes('\x1b[31m',    'ansi', 'style.bold', 'style.underline', 'style.italic', 'codes'),
        scopes('red fg, ',    'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1'),
        scopes('\x1b[44m',    'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'codes'),
        scopes('blue bg',     'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4'),
        scopes('\x1b[0m',     'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4', 'codes'),
        scopes(', normal',    'ansi')
      ]
    },

    {
      title: 'renders multi-attribute sequences',
      lines: 'normal, \x1b[1;4;3;31;44mbold, underline, italic, red fg and blue bg\x1b[0m, normal',
      expected: [
        scopes('normal, ', 'ansi'),
        scopes('\x1b[1;4;3;31;44m', 'ansi', 'codes'),
        scopes(
          'bold, underline, italic, red fg and blue bg',
          'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4'
        ),
        scopes(
          '\x1b[0m',
          'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4', 'codes'
        ),
        scopes(', normal', 'ansi')
      ]
    },
    {
      title: 'renders multi-attribute sequences ending with a semi colon',
      lines: 'normal, \x1b[1;4;3;31;44;mbold, underline, italic, red fg and blue bg\x1b[0m, normal',
      expected: [
        scopes('normal, ', 'ansi'),
        scopes('\x1b[1;4;3;31;44;m', 'ansi', 'codes'),
        scopes(
          'bold, underline, italic, red fg and blue bg',
          'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4'
        ),
        scopes(
          '\x1b[0m',
          'ansi', 'style.bold', 'style.underline', 'style.italic', 'color.fg-1', 'color.bg-4', 'codes'
        ),
        scopes(', normal', 'ansi')
      ]
    },
    {
      title: 'renders multi-attribute sequences and reverts it immediately',
      lines: 'normal, \x1b[1;4;3;31;44;0mbold, underline, italic, red fg and blue bg\x1b[0m, normal',
      expected: [
        scopes('normal, ',                                    'ansi'),
        scopes('\x1b[1;4;3;31;44;0m',                         'ansi', 'codes'),
        scopes('bold, underline, italic, red fg and blue bg', 'ansi'),
        scopes('\x1b[0m',                                     'ansi', 'codes'),
        scopes(', normal',                                    'ansi')
      ]
    },

    {
      title: 'reset fg color',
      lines: '\x1b[30mblack\x1b[39mdefault',
      expected: [
        scopes('\x1b[30m', 'ansi', 'codes'),
        scopes('black',    'ansi', 'color.fg-0'),
        scopes('\x1b[39m', 'ansi', 'color.fg-0', 'codes'),
        scopes('default',  'ansi')
      ]
    },
    {
      title: 'reset bg color',
      lines: '\x1b[40mblack\x1b[49mdefault',
      expected: [
        scopes('\x1b[40m', 'ansi', 'codes'),
        scopes('black',    'ansi', 'color.bg-0'),
        scopes('\x1b[49m', 'ansi', 'color.bg-0', 'codes'),
        scopes('default',  'ansi')
      ]
    },
    {
      title: 'resets bg and fg color in order',
      lines: '\x1b[31;44mred fg & blue bg\x1b[49mred fg\x1b[39mdefault',
      expected: [
        scopes('\x1b[31;44m',      'ansi', 'codes'),
        scopes('red fg & blue bg', 'ansi', 'color.fg-1', 'color.bg-4'),
        scopes('\x1b[49m',         'ansi', 'color.fg-1', 'color.bg-4', 'codes'),
        scopes('red fg',           'ansi', 'color.fg-1'),
        scopes('\x1b[39m',         'ansi', 'color.fg-1', 'codes'),
        scopes('default',          'ansi')
      ]
    },
    {
      title: 'resets bg and fg color in wrong order',
      lines: '\x1b[31;44mred fg & blue bg\x1b[39mblue bg\x1b[49mdefault',
      expected: [
        scopes('\x1b[31;44m',      'ansi', 'codes'),
        scopes('red fg & blue bg', 'ansi', 'color.fg-1', 'color.bg-4'),
        scopes('\x1b[39m',         'ansi', 'color.fg-1', 'color.bg-4', 'codes'),
        scopes('blue bg',          'ansi', 'color.bg-4'),
        scopes('\x1b[49m',         'ansi', 'color.bg-4', 'codes'),
        scopes('default',          'ansi')
      ]
    },

    {
      title: 'skips reset underline if not underlined',
      lines: 'not \x1b[24munderline',
      expected: [
        scopes('not ',      'ansi'),
        scopes('\x1b[24m',  'ansi', 'codes'),
        scopes('underline', 'ansi')
      ]
    },
    {
      title: 'skips reset bold if not bold',
      lines: 'not \x1b[21mbold',
      expected: [
        scopes('not ',     'ansi'),
        scopes('\x1b[21m', 'ansi', 'codes'),
        scopes('bold',     'ansi')
      ]
    },
    {
      title: 'skips reset italic if not italic',
      lines: 'not \x1b[23mitalic',
      expected: [
        scopes('not ',     'ansi'),
        scopes('\x1b[23m', 'ansi', 'codes'),
        scopes('italic',   'ansi')
      ]
    },
    {
      title: 'skips reset strike if not strike',
      lines: 'not \x1b[29mstrike',
      expected: [
        scopes('not ',     'ansi'),
        scopes('\x1b[29m', 'ansi', 'codes'),
        scopes('strike',   'ansi')
      ]
    },
    {
      title: 'skips reset blink if not blink',
      lines: 'not \x1b[25mblink',
      expected: [
        scopes('not ',     'ansi'),
        scopes('\x1b[25m', 'ansi', 'codes'),
        scopes('blink',    'ansi')
      ]
    },
    {
      title: 'skips reset hidden if not hidden',
      lines: 'not \x1b[28mhidden',
      expected: [
        scopes('not ',     'ansi'),
        scopes('\x1b[28m', 'ansi', 'codes'),
        scopes('hidden',   'ansi')
      ]
    },

    {
      title: 'ignores other escape sequences',
      lines: 'hello \x1b[Kworld',
      expected: [
        scopes('hello ', 'ansi'),
        scopes('\x1b[K', 'ansi', 'codes'),
        scopes('world',  'ansi')
      ]
    },
    {
      title: 'ignores other escape sequences with 0 parameter',
      lines: 'hello \x1b[0Kworld',
      expected: [
        scopes('hello ',  'ansi'),
        scopes('\x1b[0K', 'ansi', 'codes'),
        scopes('world',   'ansi')
      ]
    },
    {
      title: 'ignores other escape sequences with 1 parameter',
      lines: 'hello \x1b[1Kworld',
      expected: [
        scopes('hello ',  'ansi'),
        scopes('\x1b[1K', 'ansi', 'codes'),
        scopes('world',   'ansi')
      ]
    },
    {
      title: 'ignores other escape sequences with 2 parameter',
      lines: 'hello \x1b[2Kworld',
      expected: [
        scopes('hello ',  'ansi'),
        scopes('\x1b[2K', 'ansi', 'codes'),
        scopes('world',   'ansi')
      ]
    },
    {
      title: 'ignores more escape sequences',
      lines: '\x1b[40m\x1b[2J\x1b[0;34mblue \x1b[1;32mbold green\x1b[A',
      expected: [
        scopes('\x1b[40m',   'ansi', 'codes'),
        scopes('\x1b[2J',    'ansi', 'color.bg-0', 'codes'),
        scopes('\x1b[0;34m', 'ansi', 'color.bg-0', 'codes'),
        scopes('blue ',      'ansi', 'color.fg-4'),
        scopes('\x1b[1;32m', 'ansi', 'color.fg-4', 'codes'),
        scopes('bold green', 'ansi', 'style.bold', 'color.fg-2'),
        scopes('\x1b[A',     'ansi', 'style.bold', 'color.fg-2', 'codes')
      ]
    },

    {
      title: 'hides escape sequences',
      lines: '\x1b[40m\x1b[2J\x1b[0;34mblue \x1b[1;32mbold green\x1b[A\x1b[mdefault',
      hideEscapeSequences: true,
      expected: [
        scopes('\x1b[40m',   'ansi', 'codes.hidden'),
        scopes('\x1b[2J',    'ansi', 'color.bg-0', 'codes.hidden'),
        scopes('\x1b[0;34m', 'ansi', 'color.bg-0', 'codes.hidden'),
        scopes('blue ',      'ansi', 'color.fg-4'),
        scopes('\x1b[1;32m', 'ansi', 'color.fg-4', 'codes.hidden'),
        scopes('bold green', 'ansi', 'style.bold', 'color.fg-2'),
        scopes('\x1b[A',     'ansi', 'style.bold', 'color.fg-2', 'codes.hidden'),
        scopes('\x1b[m',     'ansi', 'style.bold', 'color.fg-2', 'codes.hidden'),
        scopes('default',    'ansi')
      ]
    }
  ]

  describe('tokenizeLines', () => {
    afterEach(() => {
      atom.config.set('language-ansi-styles.hideEscapeSequences', false)
    })

    testCases.forEach(({ title, lines, expected, hideEscapeSequences }) => {
      it(title, () => {
        atom.config.set('language-ansi-styles.hideEscapeSequences', !!hideEscapeSequences)
        const oneLiner = !Array.isArray(lines)
        const text = oneLiner ? lines : lines.join(os.EOL)
        expect(grammar.tokenizeLines(text)).toEqual(oneLiner ? [expected] : expected)
      })
    })
  })
})


lib/ansi-grammar.js
'use babel'

import { convert } from './convert'

export default class AnsiGrammar {
  constructor (registry) {
    this.scopeName = 'text.ansi'
    this.name = 'Ansi'
    this.fileTypes = []
    this.registry = registry
  }

  grammarUpdated () {}

  dispose () {
  }

  onDidUpdate (callback) {
    return { dispose () { } }
  }

  scopeForId (id) {
    return this.registry.scopeForId(id)
  }
  startIdForScope (scope) {
    return this.registry.startIdForScope(scope)
  }
  endIdForScope (scope) {
    return this.registry.endIdForScope(scope)
  }

  scopesFromStack (stack, rule, endPatternMatch) {
    console.log('scopesFromStack', ...arguments)
  }

  tokenizeLines (text) {
    const lines = text.split('\n')
    const scopes = []
    let ruleStack, tags
    return lines.map((line, i) => {
      ({ tags, ruleStack } = this.tokenizeLine(line, ruleStack, i === 0))
      return this.registry.decodeTokens(line, tags, scopes)
    })
  }

  tokenizeLine (line, ruleStack, firstLine) {
    return convert(line, ruleStack, firstLine, this.registry)
  }
}


lib/ansi-styles-grammar.js
'use babel'

import { convert } from './convert'

export default class AnsiStylesGrammar {
  constructor (registry) {
    this.scopeName = 'text.ansi'
    this.name = 'Ansi Styles'
    this.fileTypes = ['log', 'ansi']
    this.registry = registry
    this.maxLineLength = 10000
    this.maxTokensPerLine = 1000
  }

  grammarUpdated () {}

  dispose () {
  }

  onDidUpdate (callback) {
    return { dispose () { } }
  }

  scopeForId (id) {
    return this.registry.textmateRegistry.scopeForId(id)
  }
  startIdForScope (scope) {
    return this.registry.textmateRegistry.startIdForScope(scope)
  }
  endIdForScope (scope) {
    return this.registry.textmateRegistry.endIdForScope(scope)
  }

  scopesFromStack (stack, rule, endPatternMatch) {
    console.log('scopesFromStack', ...arguments)
  }

  tokenizeLines (text) {
    const lines = text.split('\n')
    const scopes = []
    let ruleStack, tags
    return lines.map((line, i) => {
      ({ tags, ruleStack } = this.tokenizeLine(line, ruleStack, i === 0))
      return this.registry.textmateRegistry.decodeTokens(line, tags, scopes)
    })
  }

  tokenizeLine (line, ruleStack, firstLine) {
    return convert(
      line,
      ruleStack,
      firstLine,
      this.registry.textmateRegistry,
      atom.config.get('language-ansi-styles.hideEscapeSequences')
    )
  }
}


lib/convert.js
'use babel'

const ANSI_REGEX = require('ansi-regex')()

export function convert (line, ruleStack, firstLine, registry, hideEscapeSequences) {
  const state = {
    registry,
    tags: [],
    openTags: {},
    colorFG: null,
    colorBG: null
  }

  if (firstLine) {
    open(state, 'ansi')
  } else {
    // apply every open scope from the prev line
    ruleStack.forEach(({ code, scope, codes }) => {
      if (code) {
        processCode(state, code, codes || [])
      } else if (scope) {
        open(state, scope)
      }
    })
  }

  let text = line
  const matches = line.match(ANSI_REGEX) || []
  matches.forEach((match) => {
    const index = text.indexOf(match)
    if (index) {
      // add text up to the codes
      state.tags.push(index)
      text = text.substring(index)
    }

    const len = match.length
    text = text.substring(len)

    // add tags for the codes
    const scope = 'codes' + (hideEscapeSequences ? '.hidden' : '')
    open(state, scope)
    state.tags.push(len)
    close(state, scope)

    if (!match.endsWith('m')) {
      return
    }

    // process all codes
    let codes = match.substring(2, len - 1)
    if (codes.endsWith(';')) {
      codes = codes.slice(0, -1)
    }
    codes = codes.split(';')
    while (codes.length) {
      const code = parseInt(codes.shift() || '0', 10)
      const newCodes = processCode(state, code, codes)
      if (newCodes) {
        codes = newCodes
      }
    }
  })

  // add the rest of the text
  if (text.length) {
    state.tags.push(text.length)
  }

  const newRuleStack = [
    { scope: 'ansi' }
  ]

  // carry all open tags over to the next line via the rule stack
  findOpenTags(state.tags, registry.startIdForScope('ansi'))
    .forEach((tag) => {
      newRuleStack.push(state.openTags[tag])
    })

  // atom already keeps track of the already opened scopes
  // from the previous line, so remove everything that
  // came in via the rule stack
  if (ruleStack) {
    state.tags = state.tags.slice(ruleStack.length)
  }

  return {
    line,
    tags: state.tags,
    ruleStack: newRuleStack
  }
}

const codes = {
  0: (state, code) => {
    // reset = close all existing tags
    findOpenTags(state.tags, state.registry.startIdForScope('ansi'))
      .reverse()
      .forEach((tag) => {
        close(state, state.registry.scopeForId(tag))
      })
  },
  1: (state, code) => {
    openClosable(state, code, 'style.bold')
  },
  2: (state, code) => {
    openClosable(state, code, 'style.faint')
  },
  3: (state, code) => {
    openClosable(state, code, 'style.italic')
  },
  4: (state, code) => {
    openClosable(state, code, 'style.underline')
  },
  5: (state, code) => {
    openClosable(state, code, 'style.blink') // TODO: slow
  },
  6: (state, code) => {
    openClosable(state, code, 'style.blink') // TODO: fast
  },
  7: (state, code) => {
    // TODO: fg = bg and bg = fg
    // colorBG.replace(".fg.", ".bg.")
    // colorFG.replace(".bg.", ".fg.")
  },
  8: (state, code) => {
    openClosable(state, code, 'style.hidden')
  },
  9: (state, code) => {
    openClosable(state, code, 'style.strike')
  },
  10: (state, code) => {
    // TODO: default?
  },

  11: (state, code) => {
    // TODO: different fonts?
  },
  // -19

  20: (state, code) => {
    // TODO: fraktur (Gothic)
  },
  21: (state, code) => {
    safeClose(state, 'style.bold')
    // TODO: double underline?!
  },
  22: (state, code) => {
    safeClose(state, 'style.bold')
    safeClose(state, 'style.faint')
  },
  23: (state, code) => {
    safeClose(state, 'style.italic')
  },
  24: (state, code) => {
    safeClose(state, 'style.underline')
  },
  25: (state, code) => {
    safeClose(state, 'style.blink')
  },
  26: (state, code) => {
    // 'reserved'
  },
  27: (state, code) => {
    // TODO: image positive = opposite of code 7 -> fg = fg and bg = bg
    // only apply if inversed previously!
  },
  28: (state, code) => {
    safeClose(state, 'style.hidden')
  },
  29: (state, code) => {
    safeClose(state, 'style.strike')
  },

  30: (state, code) => {
    openClosable(state, code, 'color.fg-' + (code - 30))
  },
  31: 30,
  32: 30,
  33: 30,
  34: 30,
  35: 30,
  36: 30,
  37: 30,

  38: (state, code, openCodes) => {
    const next = openCodes[0]
    if (next === '5') {
      // extended extended colors
      const openTag = openClosable(state, code, 'color.fg-' + openCodes[1])
      openTag.codes = [next, openCodes[1]] // necessary to revive this code from ruleStack
      return openCodes.slice(2)
    }
    if (next === '2') {
      // extended rgb
      // TODO: how to create custom styles here in atom?
      return openCodes.slice(4)
    }
  },
  39: (state, code) => {
    // reset FG
    if (state.colorFG) {
      safeClose(state, state.colorFG)
    }
  },

  40: (state, code) => {
    openClosable(state, code, 'color.bg-' + (code - 40))
  },
  41: 40,
  42: 40,
  43: 40,
  44: 40,
  45: 40,
  46: 40,
  47: 40,

  48: (state, code, openCodes) => {
    const next = openCodes[0]
    if (next === '5') {
      // extended extended colors
      const openTag = openClosable(state, code, 'color.bg-' + openCodes[1])
      openTag.codes = [next, openCodes[1]] // necessary to revive this code from ruleStack
      return openCodes.slice(2)
    }
    if (next === '2') {
      // extended rgb
      // TODO: how to create custom styles here in atom?
      return openCodes.slice(4)
    }
  },
  49: (state, code) => {
    // reset BG
    if (state.colorBG) {
      safeClose(state, state.colorBG)
    }
  },
  50: (state, code) => {
    // 'reserved'
  },
  51: (state, code) => {
    // TODO: framed
  },
  52: (state, code) => {
    // TODO: encircled
  },
  53: (state, code) => {
    // TODO: overlined
  },
  54: (state, code) => {
    // TODO: not framed or circled
  },
  55: (state, code) => {
    // TODO: not overlined
  },
  56: (state, code) => {
    // 'reserved'
  },
  // -59

  // 60-65: hardly ever supported
  // 60: Ideogram Underline/Right Side Line
  // 61: Ideogram Double Underline/Double Right Side Line
  // 62: Ideogram Overline/Left Side Line
  // 63: Ideogram Double Overline/Double Left Side Line
  // 64: Ideogram Stress Marking
  // 65: No Ideogram Attributes

  // 65-89: nothing

  90: (state, code) => {
    openClosable(state, code, 'color.fg-' + (code - 90 + 8))
  },
  91: 90,
  92: 90,
  93: 90,
  94: 90,
  95: 90,
  96: 90,
  97: 90,
  // 98, 99: nothing
  100: (state, code) => {
    openClosable(state, code, 'color.bg-' + (code - 100 + 8))
  },
  101: 100,
  102: 100,
  103: 100,
  104: 100,
  105: 100,
  106: 100,
  107: 100
}

function processCode (state, code, openCodes) {
  let fn = codes[code]
  if (typeof fn === 'number') {
    fn = codes[fn]
  }
  return fn ? fn(state, code, openCodes) : null
}

function open (state, scope) {
  const tag = state.registry.startIdForScope(scope)
  state.tags.push(tag)
  return tag
}

function openClosable (state, code, scope) {
  // this method is used to open style/color related scopes
  // as they are intended to be closed by other codes like "0"

  if (scope.startsWith('color.fg-')) {
    if (state.colorFG) {
      safeClose(state, state.colorFG)
    }
    state.colorFG = scope
  } else if (scope.startsWith('color.bg-')) {
    if (state.colorBG) {
      safeClose(state, state.colorBG)
    }
    state.colorBG = scope
  }

  const tag = open(state, scope)
  state.openTags[tag] = { scope, code }
  return state.openTags[tag]
}

function close (state, scope) {
  const tag = state.registry.endIdForScope(scope)
  state.tags.push(tag)
  delete state.openTags[state.registry.startIdForScope(scope)]

  if (scope.startsWith('color.fg-')) {
    state.colorFG = null
  } else if (scope.startsWith('color.bg-')) {
    state.colorBG = null
  }
}

function safeClose (state, scope) {
  const ignoreTagsForClose = [
    state.registry.startIdForScope('ansi'),
    state.registry.startIdForScope('codes'),
    state.registry.startIdForScope('codes.hidden')
  ]
  const currentOpenTags = findOpenTags(state.tags, ...ignoreTagsForClose).reverse()
  const startTag = state.registry.startIdForScope(scope)
  if (currentOpenTags[0] === startTag) {
    // easy, just close it
    close(state, scope)
  } else {
    // closing somewhere in the middle
    // close all tags up to the one we want to close
    // and open them afterwards again (in correct order of course...)
    const index = currentOpenTags.indexOf(startTag)
    if (index === -1) {
      return // not even in the list, skip it!
    }
    const toCloseAndOpen = currentOpenTags.slice(0, index).map((tag) => state.openTags[tag])
    toCloseAndOpen.forEach((o) => close(state, o.scope))
    close(state, scope)
    toCloseAndOpen.forEach((o) => processCode(state, o.code))
  }
}

function findOpenTags (tags, ...ignore) {
  const openTags = []
  tags.forEach((tag) => {
    if (ignore.indexOf(tag) !== -1) {
      return // ignored
    }
    if (tag % 2 === -1) {
      openTags.push(tag) // open
    } else {
      const index = openTags.indexOf(tag + 1)
      if (index !== -1) {
        openTags.splice(index, 1)
      }
    }
  })
  return openTags
}
